### PR TITLE
i#3044 SVE encoder: Add support for sve_int_bin_pred_log enc group.

### PR DESCRIPTION
**Pull request by fhahn**
This patch adds a new register type for SVE predicate registers
(P0-P15) and support for the sve_int_bin_pred_log encoding group.

The macros added follow the SVE assembly syntax, where the shared input
and output register Zdn has to be supplied twice.

Issue #3044

### CODE REVIEW

**Comment by fhahn** [comment]
cc @andreipoe

File: core/arch/aarch64/codec.c [hunk 1]
```diff
@@ -909,6 +909,21 @@ encode_opnd_float_reg(int pos, opnd_t opnd, OUT uint *enc_out)
     return true;
 }
 
+/* p: used to encode an SVE predicate register. */
```
Comment (derekbruening) [inline core/arch/aarch64/codec.c]: "p:" is confusing: `predicate ("p")`?

File: core/arch/aarch64/codec.c [hunk 2]
```diff
@@ -1388,6 +1403,21 @@ encode_opnd_extam(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out
     return true;
 }
 
+/* p10_low: P register at bit position 10; P0-P7 */
```
Comment (derekbruening) [inline core/arch/aarch64/codec.c]: nit: `p` in comment above but `P` here

File: core/arch/aarch64/instr_create.h [hunk 3]
```diff
@@ -1818,5 +1818,59 @@
 #define INSTR_CREATE_fnmsub_scalar(dc, Rd, Rm, Rn, Ra) \
     instr_create_1dst_3src(dc, OP_fnmsub, Rd, Rm, Rn, Ra)
 
+/* -------- SVE bitwise logical operations (predicated) ---------------- */
+
+/**
+ * Creates a ORR scalable vector instruction.
```
Comment (derekbruening) [inline core/arch/aarch64/instr_create.h]: s/a/an/

File: core/arch/aarch64/instr_create.h [hunk 4]
```diff
@@ -1818,5 +1818,59 @@
...
+ * \param Zd_     The first input SVE vector register. Must match Zd.
+ * \param Zm      The second input SVE vector register.
+ * \param width   The vector element width. Use either OPND_CREATE_BYTE(),
+ *                OPND_CREATE_HALF(), OPND_CREATE_SINGLE() or OPND_CREATE_DOUBLE().
+ */
+#define INSTR_CREATE_orr_sve_pred(dc, Zd, Pg, Zd_, Zm, width) \
+    instr_create_1dst_4src(dc, OP_orr, Zd, Pg, Zd_, Zm, width)
+
+/**
+ * Creates a EOR scalable vector instruction.
```
Comment (derekbruening) [inline core/arch/aarch64/instr_create.h]: Ditto -- and more below

File: core/arch/opnd.h [hunk 5]
```diff
@@ -272,6 +272,11 @@ enum {
     DR_REG_Z20, DR_REG_Z21,  DR_REG_Z22,  DR_REG_Z23,
     DR_REG_Z24, DR_REG_Z25,  DR_REG_Z26,  DR_REG_Z27,
     DR_REG_Z28, DR_REG_Z29,  DR_REG_Z30,  DR_REG_Z31,
+
+    DR_REG_P0,  DR_REG_P1,   DR_REG_P2,   DR_REG_P3,
```
Comment (derekbruening) [inline core/arch/opnd.h]: This breaks binary compatibility as it changes the value of  `DR_REG_NZCV`, etc. below.  (Those `DR_REG_Z*` are also newly inserted (right?) and also break it.)  Please either insert at the end or list this change under the compatibility-breaking part of the changelist in release.dox.

**Comment by fhahn** [comment]
Thanks Derek. I've addressed the comments. What do you mean with 'changelist change'?

**Comment by derekbruening** [comment]
> What do you mean with 'changelist change'?

If you went the route of breaking compatibility and editing the changelist I just wanted to take a look at that edit before merging.

File: core/arch/aarch64/encode.c [hunk 6]
```diff
@@ -79,6 +79,8 @@ const char * const reg_names[] = {
     "z8", "z9", "z10", "z11", "z12", "z13", "z14", "z15",
     "z16", "z17", "z18", "z19", "z20", "z21", "z22", "z23",
     "z24", "z25", "z26", "z27", "z28", "z29", "z30", "z31",
+    "p0", "p1", "p2", "p3", "p4", "p5", "p6", "p7",
+    "p8", "p9", "p10", "p11", "p12", "p13", "p14", "p15",
```
Comment (derekbruening) [inline core/arch/aarch64/encode.c]: This order no longer matches the enum

Comment (fhahn) [reply core/arch/aarch64/encode.c]: Done thanks. I've moved everything to the end.

**Pull request merged**